src/query/functions/temporal/aggregation.go

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package temporal
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/m3db/m3/src/query/executor/transform"
+)
+
+const (
+	// AvgTemporalType calculates the average of all values in the specified interval
+	AvgTemporalType = "avg_over_time"
+
+	// CountTemporalType calculates count of all values in the specified interval
+	CountTemporalType = "count_over_time"
+
+	// MinTemporalType calculates the minimum of all values in the specified interval
+	MinTemporalType = "min_over_time"
+
+	// MaxTemporalType calculates the maximum of all values in the specified interval
+	MaxTemporalType = "max_over_time"
+
+	// SumTemporalType calculates the sum of all values in the specified interval
+	SumTemporalType = "sum_over_time"
+
+	// StdDevTemporalType calculates the standard deviation of all values in the specified interval
+	StdDevTemporalType = "stddev_over_time"
+
+	// StdVarTemporalType calculates the standard variance of all values in the specified interval
+	StdVarTemporalType = "stdvar_over_time"
+)
+
+var (
+	aggFuncs = map[string]interface{}{
+		AvgTemporalType:    struct{}{},
+		CountTemporalType:  struct{}{},
+		MinTemporalType:    struct{}{},
+		MaxTemporalType:    struct{}{},
+		SumTemporalType:    struct{}{},
+		StdDevTemporalType: struct{}{},
+		StdVarTemporalType: struct{}{},
+	}
+)
+
+// NewAggOp creates a new base temporal transform with a specified node
+func NewAggOp(args []interface{}, optype string) (transform.Params, error) {
```

Comment:
I'd recommend just taking floats or whatever as your args, and have the logic for providing usable values in the parser instead, so you don't end up with extra parsing scattered around in the function op.

Reply:
i think the problem here is that some ops require more than one args. Then the question is where should that logic live which determines which argument is need by which function.

Reply:
We should add a todo to consolidate how we pass args around, https://github.com/m3db/m3/pull/884 should help us standardize scalar parameters, but unfortunately there's a few special cases in prom which take `Scalar, Fetch` instead of `Fetch, Scalar`, or even `Fetch, Scalar, Scalar`. Prom have a set of required param types by position; may be an idea to do something similar?